### ISSUE TITLE
FunASR-nano 的 ONNX 性能与原版有较大差距

### ISSUE THREAD

**Issue by HaujetZhao** [DakeQQ/Automatic-Speech-Recognition-ASR-ONNX#16]
我用测试转录同一段音频，130秒的音频。

运行 FunASR 原版，载入模型后，任务管理器显示 GPU 显存已占用2.4G，在推理的时候，占用到3.6G，最后每段的 RTF 能达到 0.1 左右，耗时13秒转录完成。

而用导出的 ONNX 模型去解码的话，显存占用达到了5.9GB，我用了更短的音频去进行转录，耗时23秒。所需耗时是原版的 1.8 倍。

尝试了一天让 Gemini 辅助方案，用 Claude Code 写，但我还是太菜，最后额度用完了，空忙活了一天也没效果。只能来 issue 告知一下了。

以下是 Gemini 的分析，供参考。

---

这**完全不是**绝望的境地，而是因为目前的导出脚本采用了“最朴素”的导出方式（FP32精度 + 未融合算子），导致了显存翻倍和计算效率低下的问题。

PyTorch 原版快是因为：
1.  它默认可能在用 BF16/FP16 运行（Qwen 的默认精度）。
2.  PyTorch 内部使用了 FlashAttention 等融合算子，极大地优化了显存访问。

目前的 ONNX 导出脚本存在以下三个**致命瓶颈**：
1.  **强制 FP32**：代码中显式调用了 `.float()`，导致模型权重和 KV Cache 全部使用 32位浮点数。这直接导致显存占用比 PyTorch（通常是 FP16/BF16）翻倍，内存带宽压力也翻倍。
2.  **KV Cache 拷贝机制**：目前的 Decoder 写法是 `torch.cat([old, new])` 并输出完整的 KV Cache。这意味着每一步生成，ONNX Runtime 都要把越来越大的 KV Cache 在内存中搬运一遍。
3.  **算子未融合**：导出的 ONNX 是由大量基础算子（MatMul, Add, Softmax）组成的，没有使用 ONNX Runtime 的 `Attention` 或 `MultiHeadAttention` 融合算子，效率远低于 PyTorch 的实现。

---

### 解决方案：使用 ORT Transformers Optimizer 进行 FP16 转换与算子融合

你不需要重新重写复杂的导出逻辑，只需要在导出 FP32 模型后，加一个**后处理步骤**。

请创建一个新的 Python 脚本（例如 `optimize_onnx.py`），运行它来优化你导出的模型。

#### 1. 优化脚本 (`optimize_onnx.py`)

你需要安装 `onnxruntime-tools` 或者直接使用 `onnxruntime` 内置的工具（建议安装 `pip install onnxruntime-gpu`）。

```python
import onnx
import os
from onnxruntime.transformers import optimizer
from onnxruntime.transformers.fusion_options import FusionOptions

# 你的模型路径
model_dir = './Fun-ASR-Nano-2512-onnx'
output_dir = './Fun-ASR-Nano-2512-onnx-fp16'

if not os.path.exists(output_dir):
    os.makedirs(output_dir)

# 需要优化的模型列表
# 注意：Encoder 和 Decoder 是 Transformer 结构，收益最大
# Beam Search 等逻辑控制模型不需要 FP16 转化，保持原样即可或者仅做简单优化
models_to_optimize = [
    ('FunASR_Nano_Encoder.onnx', 'bert'),  # Encoder 结构类似 BERT/Transformer-Encoder
    ('FunASR_Nano_Decoder_Main.onnx', 'gpt2'), # Decoder 结构类似 GPT
    ('FunASR_Nano_Decoder_Embed.onnx', 'no_opt'), # Embedding 层通常保持原样或仅FP16
]

# 仅仅复制不需要重度优化的模型
import shutil
for f in ['FunASR_Nano_Greedy_Search.onnx', 'FunASR_Nano_First_Beam_Search.onnx', 
          'FunASR_Nano_Second_Beam_Search.onnx', 'FunASR_Nano_Reset_Penality.onnx']:
    src = os.path.join(model_dir, f)
    dst = os.path.join(output_dir, f)
    if os.path.exists(src):
        shutil.copy(src, dst)
        print(f"Copied {f}")

for model_name, model_type in models_to_optimize:
    input_path = os.path.join(model_dir, model_name)
    output_path = os.path.join(output_dir, model_name)
    
    print(f"Optimizing {model_name}...")
    
    if model_type == 'no_opt':
        # 仅转换 FP16，不融合算子（Embed层）
        from onnxruntime.transformers.onnx_model import OnnxModel
        model = onnx.load(input_path)
        onnx_model = OnnxModel(model)
        onnx_model.convert_float_to_float16()
        onnx_model.save_model_to_file(output_path)
        continue

    # 配置融合选项
    optimization_options = FusionOptions(model_type)
    optimization_options.enable_gelu_approximation = True # 允许 GELU 近似，提速
    
    # 运行优化器
    # num_heads 和 hidden_size 需要根据你的 Qwen 0.6B 配置填写
    # Qwen 0.6B: hidden_size=896 (可能), num_heads=14 (需确认 model.py 里的 config)
    # 如果不知道确切参数，use_gpu=True 通常能自动推断
    opt_model = optimizer.optimize_model(
        input_path,
        model_type=model_type, 
        num_heads=0, # 设为0让优化器自动推断
        hidden_size=0, # 设为0让优化器自动推断
        use_gpu=True,
        opt_level=1 # 1: Basic, 2: Extended (可能会改变精度，先试1或2)
    )
    
    # 转换为 FP16
    opt_model.convert_float_to_float16()
    
    # 保存
    opt_model.save_model_to_file(output_path)
    print(f"Saved optimized FP16 model to {output_path}")

print("\nOptimization Complete! update your inference script to point to the new folder.")
```

#### 2. 修改推理脚本 (`run_inference.py`)

优化后的模型输入输出类型变为了 `float16`。你需要修改推理脚本中的数据类型。

**修改点 1：模型路径**
指向新的 `Fun-ASR-Nano-2512-onnx-fp16` 文件夹。

**修改点 2：数据类型 (Critical)**
在 `run_inference.py` 中，你需要确保输入给 Decoder 的 Tensor 是 FP16 的。

```python
# 在 main 函数开头附近
# ...
# 强制使用 float16，因为我们转换了模型
model_dtype = np.float16 

# ...

# 所有的初始化 Tensor 都要改成 model_dtype
penality_value = onnxruntime.OrtValue.ortvalue_from_numpy(np.array([REPEAT_PENALITY], dtype=model_dtype), device_type, DEVICE_ID)
init_past_keys_C = onnxruntime.OrtValue.ortvalue_from_numpy(np.zeros(..., dtype=model_dtype), device_type, DEVICE_ID)
init_past_values_C = onnxruntime.OrtValue.ortvalue_from_numpy(np.zeros(..., dtype=model_dtype), device_type, DEVICE_ID)
init_repeat_penality = onnxruntime.OrtValue.ortvalue_from_numpy(np.ones(..., dtype=model_dtype), device_type, DEVICE_ID)

# ...

# 检查 Encoder 的输入
# Encoder 的输入 audio 还是 int16/float32 没关系，因为第一层通常会处理
# 但如果 Encoder 也转了 FP16，可能需要检查 query_embed 是否需要转 float16
# 原代码：prompt_embed = ort_session_B.run_with_ort_values(...)[0]
# 这里的 Embed 输出自然会变成 float16 (因为模型B被转了)，所以后续传给 Decoder C 应该没问题。
```

#### 3. 为什么这样做能解决问题？

1.  **显存减半**：FP16 权重大小是 FP32 的一半。KV Cache 的大小也是一半。你的显存占用应该会从 5.9GB 降到 3GB 左右（接近 PyTorch 水平）。
2.  **速度提升**：
    *   **Tensor Core 利用**：NVIDIA GPU 在 FP16 下的计算能力（TFLOPS）远高于 FP32。
    *   **带宽节省**：内存读写量减半。
    *   **算子融合**：`optimize_model` 会把 `LayerNorm + MatMul + Gelu` 等操作合并成一个大算子，减少 GPU 启动 kernel 的开销。

#### 4. 进阶优化（如果上述还不够）

如果 FP16 后速度还不够（比如 RTF 还没到 0.1），问题在于 `Export_Fun_ASR_Nano.py` 中的 KV Cache 导出逻辑。

目前的逻辑：
`k = torch.cat((all_inputs[i], k), dim=-1)`

在 ONNX 中，这会产生大量显存碎片的分配和拷贝。

**理想的 LLM ONNX 结构（Static KV Cache）：**
这需要修改导出脚本 `Export_Fun_ASR_Nano.py` 中的 `FUNASR_NANO_DECODER_MAIN`。
*   不使用 `torch.cat` 并返回新的大 Tensor。
*   改为输入固定大小的 `past_key_values`（例如最大长度 1024）。
*   使用 `ScatterND` 或者 `index_put_` 仅仅更新当前 token 对应的 KV 位置。
*   模型输出不包含 KV Cache，或者只输出新增的那一列。

但鉴于修改导出逻辑比较复杂，**强烈建议先尝试上面的 FP16 + Optimizer 方案**。这通常能带来 2-3 倍的提速和 50% 的显存节省，足以让 ONNX 版本达到甚至超过 PyTorch 原版的性能（PyTorch Eager mode 其实并不快，ONNX Runtime 优化后通常更快）。

### 总结
请先运行上述的 `optimize_onnx.py` 脚本，将模型转换为 FP16 并开启融合。不要在 FP32 的原始导出模型上死磕，那是注定效率低下的。

**Comment by DakeQQ** [comment]
关于 llama.cpp、Ollama 或 vLLM 的建议很有参考价值，在桌面端上确实能发挥更强的性能。但我們目前聚焦于移动端及嵌入式平台赛道，而 ONNX Runtime 在这些特定硬件上的适配性更具优势与便利性。精力有限，我们暂时没能分身提供其他框架的部署范例，还请您见谅。

**Comment by DakeQQ** [comment]
牛阿～
您这GGUF是不是有量化压缩过？

**Comment by HaujetZhao** [comment]
导出来的 gguf 是 1.11GB，应该是 fp16 的，原版在运行的时候应该也是 fp16

**Comment by HaujetZhao** [comment]
完整流程已上传到   https://github.com/HaujetZhao/FunASR-nano-onnx-gguf

**Comment by HaujetZhao** [comment]
<img width="2880" height="1800" alt="Image" src="https://github.com/user-attachments/assets/72d387c4-42c1-422e-b999-aba8492d83ce" />

这就是用 CPU 跑 int8 的极限速度了，看起来 RTF 偏低，实则是因为音频文件中的语音特别密集，文字特别多。

或许编码器速度还能优化，但我没能力。

[input.mp3](https://github.com/user-attachments/files/24756897/input.mp3)

**Comment by HaujetZhao** [comment]
<img width="2880" height="1800" alt="Image" src="https://github.com/user-attachments/assets/c217981b-b63d-4d8a-bb59-70e90984259c" />

<img width="2880" height="1800" alt="Image" src="https://github.com/user-attachments/assets/296f6f12-5e5f-41f5-aea8-3f42120a3be1" />

Decoder 的量化加速也搞定了，速度爽到飞起！！！

**Comment by qgzang** [comment]
跟两位大佬学习，急需一个加速版本的FunASR-nano。

**Comment by DakeQQ** [comment]
经 SOTA Agent 优化后，現在所有 ASR 模型达到了原 PyTorch 推理的精度，还发挥出了 ONNX Runtime 的最大性能。

**Comment by HaujetZhao** [comment]
> 经 SOTA Agent 优化后，現在所有 ASR 模型达到了原 PyTorch 推理的精度，还发挥出了 ONNX Runtime 的最大性能。

我持怀疑态度。看了下 fun nano 和 firered 的导出代码，都用的是 dynamo  = False ，单凭这点，就可以说还没到最好性能。

默认情况下，大部分模型官方定义，都不能 dynamo 导出，但是做算子的改写后就可以，能提升不小的速度表现。

**Comment by HaujetZhao** [comment]
我这里有一些 dynamo 导出的经验文档，可以让 agent 看后进行改造，应该能成功用上 dynamo：

https://github.com/HaujetZhao/Fun-ASR-GGUF/tree/master/experience

https://github.com/HaujetZhao/Qwen3-TTS-GGUF/tree/main/Experience

**Comment by DakeQQ** [comment]
目前测试发现，Fun-ASR-Nano + Dynamo + Q8 + CPU 未能带来明显正向收益。以下的测试数据可以参考参考。

---

# Fun-ASR-Nano Q8：Legacy 与 Dynamo 对比

| 项目 | Legacy Q8 | Dynamo Q8 | 结论 |
|---|---:|---:|---|
| 转写正确性 | 短音频 LLM/CTC：16 / 16 通过；长音频：通过 | 短音频 LLM/CTC：16 / 16 通过；长音频：通过 | **功能一致** |
| Q8 部署体积 | 911 MB | 911 MB | **无体积差异** |
| 短音频 RTF，4 轮平均 | 0.14563 | 0.14319 | 差异在 CPU 波动范围内，无明确胜负 |
| 长音频 RTF，122 秒、5 个窗口、post-warm | **0.09867** | 0.10200 | **Dynamo 高 3.4%，Legacy 略快** |
| 长音频 decode 速度 | 29.130 token/s | 29.017 token/s | 基本相同 |
| 长音频 prefill 时间 | **6.314 s** | 6.688 s | Dynamo 高 5.9%，是主要差异来源 |
| 图节点数，Prefill Penalty Greedy | 3,101 | 3,123 | Dynamo 多 22 个动态 shape/indexing 节点 |
| 图节点数，Decode Penalty Greedy | 1,280 | 1,289 | Dynamo 多 9 个动态 shape/indexing 节点 |
| 图结构差异 | 较精简 | 多出 `Shape`、`Reshape`、`Concat`、`GatherND` 等控制节点 | 不改变功能或体积，但增加 prefill 控制开销 |

---

**Comment by HaujetZhao** [comment]
我试了你的 Firered ，也做了 dynamo 尝试，确实，你做的算子优化已经很好，已经性能最佳，dynamo 只消除了一些 cast 算子，没有多大提升。
